layouts: simple confirmations and warnings

mmilata opened the pull request:
This generalizes the `confirm_action` layout a bit and most of the relevant application code is modified to use it. Same with `show_success`/`show_warning`/`show_error`.

[UI diff](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/1033168650/artifacts/master_diff/index.html)

andrewkozlik commented:
Looking at the UI changes, I can see that the rendering mechanism adds a space after each parameter, e.g. `after 10 seconds?` is now `after 10 seconds ?`, which is incorrect. Hmm, I am guessing that this is due to the fact that `format_parametrized()` strips the spaces and `render_text()` puts them back, which is wrong if no space was stripped. So why do we strip them in the first place?

matejcik commented:
> why do we strip them in the first place?

because otherwise there will be two spaces.
we can modify `render_text` to not add a space, but that entails reviewing all places that rely on this behavior, which is a lot IIRC

andrewkozlik commented:
> > why do we strip them in the first place?
> 
> because otherwise there will be two spaces.
> we can modify `render_text` to not add a space, but that entails reviewing all places that rely on this behavior, which is a lot IIRC

So let me rephrase the question. Why do we add the spaces? Do we have any examples that rely on this behavior?

mmilata commented:
I don't know the original reasoning but at this point I'd say we add the spaces in `render_text` to stay compatible with the previous implementation. Given `text = Text("title"); text.normal("foo", "bar")`, there is going to be either linebreak or space between the two words. As a result there's no way to render two characters of different font next to each other, which would be the right thing to do here.

We probably could change the behavior and update all invocations to be like `text.normal("foo ", "bar")`, however I'd propose to instead solve this by making the `?` part of the parameter, which:
* makes the screen identical to what we currently have in master, and
* does not make the ongoing Rust rewrite of the `render_text` harder.

andrewkozlik commented:
> We probably could change the behavior and update all invocations to be like `text.normal("foo ", "bar")`, however I'd propose to instead solve this by making the `?` part of the parameter (and thus bold), which:
> 
> * makes the screen identical to what we currently have in master, and
> 
> * does not make the ongoing Rust rewrite of the `render_text` harder.

It seems to me that the correct solution would be not to strip spaces and not to add them at all. That way, if `new_lines=True` then `text.normal("foo", "bar")` remains as is and if `new_lines=False`, then it changes to `text.normal("foo bar")`. I don't think we need to worry about making the screens identical to what we currently have in master, because it's not always correct. Like the question mark, for example, really shouldn't be bold. I can't judge the impact on the ongoing Rust rewrite, but not sure if that should be a factor.

mmilata commented:
So I removed the spaces in 2febef0 and not much was needed to make UI tests pass: [UI diff from master](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/1037645711/artifacts/master_diff/index.html).

Still this change probably introduces some corner case that will break something not covered by test. The code converted to layouts mostly uses `"single strings\nwith embedded\nnewlines"` which is fine, I'll need to review the rest while converting it to layouts.

![boldquestionmark](https://user-images.githubusercontent.com/85857/108270348-9cbef180-716f-11eb-9d8f-33f98102ccdf.png)

mmilata commented:
Added two more commits to finish `apps.bitcoin` conversion. [UI diff](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/1052854587/artifacts/master_diff/index.html)

tsusanka reviewed (approved):
LGTM with the exception of the "require" usage. As discussed on standup it seems safer to have "require" by default and explicitly omit "requiredness" if necessary.

matejcik commented:
I'd prefer to look this over before it is merged. If I'm blocking something, I'll move this up my priority list :) just let me know

mmilata commented:
Mind if I rebase this to master?

matejcik commented:
> Mind if I rebase this to master?

go ahead

mmilata commented:
Rebased. [UI diff](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/1090171496/artifacts/master_diff/index.html)

Compared to master the total allocation count in device tests increased by 42488 (0.3 %). [master](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/1089743515/artifacts/core/prof/memperf-html/index.html) -> [this branch](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/1090171494/artifacts/core/prof/memperf-html/index.html)

matejcik commented on core/src/trezor/ui/layouts/common.py:
```diff
@@ -29,3 +21,11 @@ async def interact(
     return await ctx.wait(layout)
+
+
+async def not_cancelled(a: Awaitable[None]) -> bool:
```
ISTM all uses of `not_cancelled` would be superseded by passing a custom exception instance to `show_error` or `confirm_action`, correct?

mmilata replied:
Yeah, good idea:+1:. Fixed in 3c39b4ee1b842dbf9baf26e2491f4697009a1b45.

matejcik commented on core/src/trezor/ui/layouts/tt.py:
```diff
@@ -50,58 +60,105 @@
+    reverse: bool = False,  # TODO cleanup @ redesign
+    larger_vspace: bool = False,  # TODO cleanup @ redesign
     br_code: EnumTypeButtonRequestType = ButtonRequestType.Other,
     **kwargs: Any,
```
this is now unused

mmilata replied:
Fixed in 28f676ca68135ca4a6b46cdf7b18815a361e4b36.

matejcik commented on core/src/trezor/ui/layouts/tt.py:
```diff
@@ -407,4 +567,67 @@ def confirm_modify_fee(
+
+
+async def confirm_signverify(
+    ctx: wire.GenericContext, coin: str, message: str, address: str = None
```
`address: str = None` seems like an invalid type declaration since some changes in mypy. Is this covered by typecheck?

mmilata replied:
It's illegal wrt [PEP484](https://www.python.org/dev/peps/pep-0484/#union-types) but mypy ([still](https://github.com/python/mypy/issues/9091)) allows it. This behavior can be [turned off](https://github.com/trezor/trezor-firmware/blob/ba326865f1b9cacf25eb7d758ba45385cdc41fb7/setup.cfg#L48) but we rely on in in ~300 places. I'll fix the usage in this PR and will look into the rest of the codebase afterwards.

matejcik replied:
also worth considering is switching to the new-(new-)style declarations with `str | None`

matejcik replied:
you can configure mypy to assume python 3.10

mmilata replied:
Fixed in 4b27e3b0d265ae8a0d16afc78b3c3e688e1e9054.

matejcik reviewed (changes requested):
Many comments that are actually one comment: ISTM there is a better alternative to `not_cancelled` function. 

Larger comment: `confirm_action` has way too many parameters now that should go away in the future. I am assuming all the "TODO cleanup" means that you're leaving them in for now for the sake of a clean UI diff?

For the most part though, this is shaping up wonderfully.

mmilata commented on core/src/trezor/ui/layouts/tt.py:
```diff
@@ -299,102 +429,132 @@ def show_success(
+    icon: str = ui.ICON_SEND,  # TODO cleanup @ redesign
+    icon_color: int = ui.GREEN,  # TODO cleanup @ redesign
+    width: int = MONO_HEX_PER_LINE,
+    truncate: bool = False,
```
It seems to me that there usually isn't a reason to truncate the data, and if there is it probably should be done explicitly. Maybe in the future we should go as far as to make the pagination mandatory whenever a text doesn't fit single screen?

matejcik replied:
let's remove the parameter now, and switch to implicitly paginating when that part of code is ready

mmilata replied:
I'm using the parameter quite a bit in the commits that will follow this PR in order to keep the UI diff minimal. Looking into this.

mmilata replied:
Removed the parameter in 49006de124a5246d58018746cc9bb4117cbb9770, let's discuss it in the next PR with more context.

I think I addressed all your comments, PTAL.

matejcik replied:
are you ever setting it to `False` explicitly?

mmilata replied:
nope

mmilata commented on core/src/apps/management/wipe_device.py:
```diff
@@ -1,12 +1,12 @@
 
 async def wipe_device(ctx, msg):
-    await require(confirm_wipe(ctx))
+    await confirm_wipe(ctx)
```
There's some styling missing but since `confirm_action` has everything (w/ kitchen sink) anyway I fixed this in 28f676ca68135ca4a6b46cdf7b18815a361e4b36.

mmilata commented on core/src/apps/webauthn/add_resident_credential.py:
```diff
@@ -41,14 +41,16 @@ async def add_resident_credential(
-            "not belong to this",
-            "authenticator.",
+        await not_cancelled(
+            show_error(
```
Fixed in 3c39b4ee1b842dbf9baf26e2491f4697009a1b45.

mmilata commented on core/src/apps/management/recovery_device/layout.py:
```diff
@@ -23,13 +23,29 @@
 
 async def confirm_abort(ctx: wire.GenericContext, dry_run: bool = False) -> bool:
```
Fixed in 3c39b4ee1b842dbf9baf26e2491f4697009a1b45.

mmilata commented on core/src/apps/common/sdcard.py:
```diff
@@ -1,9 +1,7 @@
-from trezor.ui.components.tt.text import Text
-
-from apps.common.confirm import confirm, hold_to_confirm
+from trezor.ui.layouts import confirm_action, not_cancelled, show_error
```
Fixed in 3c39b4ee1b842dbf9baf26e2491f4697009a1b45.

mmilata commented on core/src/trezor/ui/layouts/tt.py:
```diff
@@ -250,46 +323,103 @@ async def show_address(
+    )
+
+
+async def show_error(
```
Went with `show_error_and_raise` but don't feel strongly about it. Fixed in 3c39b4ee1b842dbf9baf26e2491f4697009a1b45.

mmilata commented on core/src/apps/common/request_pin.py:
```diff
@@ -124,18 +124,30 @@ async def verify_user_pin(
-    await confirm(ctx, text, confirm=None, cancel="Close")
+    from trezor.ui.layouts import show_error, not_cancelled
+
+    await not_cancelled(
```
Fixed in 3c39b4ee1b842dbf9baf26e2491f4697009a1b45.

mmilata commented:
Thanks for the review @matejcik! Yes, the parameters with the comment are expected to disappear but exist for the sake of changing the UI as little as possible.

Any idea how to verify that the PR doesn't make our fragmentation issues worse? Checking whether individual tests pass/fail on an actual hardware is very time consuming. Perhaps we can run the tests on an emulator repeatedly, decreasing the heap size each iteration until some number of tests start to fail, then compare with master?

matejcik commented:
> Perhaps we can run the tests on an emulator repeatedly, decreasing the heap size each iteration until some number of tests start to fail, then compare with master?

this is reasonable. for starters, i'd like to get `test_15_of_15` to fail on emulator, which might give us a reasonable bound for the heap size

also i wouldn't merge this into the upcoming release, which means we will most likely get the workflow-restart thing so that fragmentation is much less of an issue

matejcik commented on core/src/apps/common/request_pin.py:
```diff
@@ -124,18 +124,28 @@ async def verify_user_pin(
+        red=True,
+        exc=wire.PinInvalid,
+    )
     raise wire.PinInvalid
```
isn't this dead code now?

mmilata replied:
It is but mypy needs it:
```
src/apps/common/request_pin.py:140: error: Implicit return in function which does not return  [misc]
```
Which is weird because `show_error_and_raise` is declared as `NoReturn`.

matejcik replied:
that's what I was thinking too. weird indeed. maybe something to do with asyncness of the function?
in any case let's put in an `assert False` instead to make it clearer for the reader

mmilata replied:
Fixed in fef81373696d5dc2929b40dba0343afe855c9a7f. `assert False` does nothing in production firmware, right?

matejcik replied:
correct

matejcik commented on core/src/apps/common/request_pin.py:
```diff
@@ -124,18 +124,28 @@ async def verify_user_pin(
+        red=True,
+        exc=wire.PinInvalid,
+    )
     raise wire.PinInvalid
```
isn't this dead code now?

mmilata replied:
Fixed in fef81373696d5dc2929b40dba0343afe855c9a7f.

matejcik commented on core/src/apps/common/sdcard.py:
```diff
@@ -13,61 +11,105 @@ class SdCardUnavailable(wire.ProcessError):
+        ctx,
+        "warning_sd_retry",
+        "SD card problem",
+        action="None",
```
didn't notice it before but this should probably be `None`, not `"None"`

matejcik replied:
also means this is not covered by UI tests, correct?
that's probably because the SD card persistence tests don't have UI recording...

mmilata replied:
Oof, fixed in 82b690518f6c60f4d81c22421c45ad3e06322ee1. Will look into writing a testcase.

mmilata replied:
There doesn't seem to be a simple way to inject SD card errors that lead to these dialogs:/ Checked manually that both work and look OK.

matejcik commented on core/src/apps/webauthn/add_resident_credential.py:
```diff
@@ -41,15 +41,15 @@ async def add_resident_credential(
+            button="Close",
+            content="The credential you are trying to import does\nnot belong to this authenticator.",
+            red=True,
+            exc=wire.ActionCancelled,
```
this is the default, no?

mmilata replied:
Fixed in fef81373696d5dc2929b40dba0343afe855c9a7f.

matejcik commented on core/src/trezor/ui/components/common/confirm.py:
```diff
@@ -1,16 +1,26 @@
+    return x is CONFIRMED
+
+
+async def raise_if_cancelled(a: Awaitable, exc: Any = wire.ActionCancelled) -> None:
```
the type should probably be `BaseException | Type[BaseException]` (here and in the other places that pass this through

mmilata replied:
`wire.ActionCancelled` is an `int` though?

matejcik replied:
i'm somewhat sure it's an `Error`

mmilata replied:
D'oh, was looking `message/FailureType.py`. Maybe I should try something more intelligent than `ctags`.:thinking:. Fixed in fef81373696d5dc2929b40dba0343afe855c9a7f.

matejcik reviewed (approved):
the `sdcard` thing is looking much neater now :heart: 

i believe this is it. but let's discuss before merging if we want this to go to the upcoming release

mmilata commented:
So I've tested this with
```
core/emu.py --disable-animation --temporary-profile --heap-size=460K &
TESTOPTS="--random-order-bucket=none" make test_emu
```
and it works but with `--heap-size=450K` it fails on `test_15_of_15`. These two bounds are the same for master so it probably doesn't make the fragmentation much worse.

matejcik commented:
let's get this merged

@mmilata please squash, rebase and fix conflicts

mmilata commented:
Rebased, squashed, tests pass. Re-ran the heap sanity test, minimum `--heap-size` value is 459K for this branch and >=460K for master. Will merge this in the evening if there are no objections.
